Implement aggregations on aggregate metric fields

csoulios opened the pull request:
This PR will merge all work done for aggregate-metric aggregations in the feature branch back into master.

----- 

In the process of developing a new implementation for the Elasticsearch Rollups functionality we came up with the concept of the aggregate metric field type.

The `aggregate_metric_double` field type can store the results of aggregations (currently min, max, sum, value_count and avg are supported - more to come).

For example, let's assume we have the following documents in an index and metric field is of aggregate_metric type:

```
{"metric": {"min": 0, "max": 100, "sum": 405, "value_count": 50} }
{"metric": {"min": 60, "max": 100, "sum": 10000.45, "value_count": 100} }
{"metric": {"min": -400.50, "max": 1000, "sum": -400.50, "value_count": 1000} }
{"metric": {"min": 1, "max": 99.3, "sum": 1, "value_count": 20} }
{"metric": {"min": -100, "max": -40, "sum": -100, "value_count": 40} }
```

This field allows us to run (min, max, sum, value_count, avg) aggregations as in the following example:

```
POST /index_agg/_search?size=0
{
  "aggs" : {
    "max_metric" : { "max" : { "field" : "metric" } }
  }
}
```

Note that the max aggregation is run on the `metric` field and not the `metric.max` sub-field. However, the field mapper will compute the results of the max aggregation based on the `metric.max` sub-field.
In fact the `aggregate_metric_double` field type will return the max sub-field to the max aggregation, the min sub-field to the min aggregation and so on.

Reading all this, one would think why is this different to having an index that stores one numeric field for each aggregation.

```
{"metric_min": 0, "metric_max": 100, "metric_sum": 405, "metric_value_count": 50}
```
and then run aggregations on those fields (eg `"max_metric" : { "max" : { "field" : "max_metric" } }` )

The power of the `aggregate_metric _double` field type comes from **its ability to mix indices with raw (numeric) fields and `aggregate_metric_double` fields in the same query**.

For example, let's assume we have one aggregate index and one index with raw data:

**index_agg:**
```
{"metric": {"min": 0, "max": 100, "sum": 405, "value_count": 50} }
{"metric": {"min": 60, "max": 100, "sum": 10000.45, "value_count": 100} }
{"metric": {"min": -400.50, "max": 1000, "sum": -400.50, "value_count": 1000} }
```

**index_raw:**
```
{"metric": 100 }
{"metric": 34 }
{"metric": -100.45}
```

We can run the max aggregation on both indices

```
POST /index_agg,index_raw/_search?size=0

{
  "aggs" : {
    "max_metric" : { "max" : { "field" : "metric" } }
  }
}
```

and the aggregation will merge the partial aggregations over the two indices.

This feature can be particularly useful for specific use cases that benefit from summary fields.

csoulios committed:
Implement aggregate_metric field mapper (#49830)

Implemented aggregate_metric field type for storing pre-computed aggregates

csoulios committed:
Merge branch 'feature/aggregate-metrics' of github.com:elastic/elasticsearch into feature/aggregate-metrics

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Merge with master - fixes

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Implement aggregations on aggregate metrics (#53986)

Following the implementation of the aggregate_metric_double field mapper(#49830) we are implementing the Min, Max, ValueCount, Sum and Average aggregations on aggregate metrics.

The code builds on the excellent work done for #42949 and uses the extensible ValuesSources infrastructure to wire up common metric aggregation on the aggregate_metric_double field type.

This PR is part of the rollups v2 refactoring as described in meta issue #42720

elasticmachine commented:
Pinging @elastic/es-analytics-geo (:Analytics/Aggregations)

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

not-napoleon commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/aggregations/metrics/AggregateMetricsAggregatorsRegistrar.java:
```diff
@@ -0,0 +1,104 @@
+            (MetricAggregatorSupplier) (name, valuesSource, formatter, context, parent, metadata) -> new AggregateMetricBackedSumAggregator(
+                name,
+                (AggregateMetricsValuesSource.AggregateDoubleMetric) valuesSource,
+                formatter,
```
Just a heads up - I changed `MetricAggregatorSupplier` to take a `ValuesSourceConfig` here instead of separate `ValuesSource` and `DocValuesFormat` args.  I've cleaned up VSConfig so it can be used as an actual config object rather than having to pick it apart in the factories.

csoulios replied:
Thanks @not-napoleon. I will fix that

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Resolve conflicts from master

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Fix integration tests

talevy committed:
Merge remote-tracking branch 'elastic/master' into feature/aggregate-metrics

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Updated branch to fix build after merge

talevy committed:
Merge remote-tracking branch 'elastic/master' into feature/aggregate-metrics

talevy committed:
update code in latest merge

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Fix errors after merge

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Fix errors after merge

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Fix errors after merge

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Fix errors after merge

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios commented:
@elasticmachine run elasticsearch-ci/1

csoulios commented:
@elasticmachine run elasticsearch-ci/1

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Converted aggregate_metric_double field mapper to parameterized mapper (#64172)

AggregateDoubleMetricFieldMapper was originally not written as a parametrized field mapper.

This PR refactors AggregateDoubleMetricFieldMapper to extend ParametrizedFieldMapper.

Also, refactored AggregateDoubleMetricFieldMapperTests to extend MapperTestCase instead of ESSingleNodeTestCase

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Fix errors after merge

talevy committed:
Merge remote-tracking branch 'elastic/master' into feature/aggregate-metrics

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics

csoulios committed:
Fix errors after merge

talevy committed:
Merge remote-tracking branch 'elastic/master' into feature/aggregate-metrics

talevy committed:
Merge remote-tracking branch 'elastic/master' into feature/aggregate-metrics

talevy committed:
Hide metric sub-fields from aggregate_metric searches and add support for scripting (#64323)

This commit only adds one searchable field to the aggregate_metric
field mapper. It makes the defaultMetric value searchable under the
same field name as the aggregate_metric.

This should make aggregate_metric behave more like a numeric in
most use-cases.

```
{
    ...
    "fields": {
        "temperature": {
            "aggregate_metric_double": {
                "type": "aggregate_metric_double",
                "searchable": true,
                "aggregatable": true
            }
        }
    }
}
```

Other things in this PR

- `default_metric` is now required and does not default to `max`
- the fields API's valueFetcher for `aggregate_metric_double` now returns the default-metric value
- support for scripting, where the doc-value returned is that of the `default_metric`

talevy committed:
Merge remote-tracking branch 'elastic/master' into feature/aggregate-metrics

csoulios committed:
Merge branch 'master' into feature/aggregate-metrics
